### PR TITLE
Upgrade Scala version from 2.13.8 to 2.13.14

### PR DESCRIPTION
**Pull request by dupen01**
# :mag: Description
## Issue References 🔗

To support Java 21, the Scala version needs to be upgraded to 2.13.11 or above.

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Related Unit Tests

---

# Checklist 📝

- [ ] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

**Be nice. Be informative.**

### CODE REVIEW

**Comment by pan3793** [comment]
can we use Scala 2.13.14 to align with Spark 4.0.0-preview1?

**Comment by pan3793** [comment]
BTW, please rebase your code on latest master branch, there were some issues on CI in my previous commits

File: pom.xml [hunk 1]
```diff
@@ -1930,7 +1930,8 @@
             <id>scala-2.13</id>
             <properties>
                 <scala.binary.version>2.13</scala.binary.version>
-                <scala.version>2.13.8</scala.version>
+                <!-- Scala version above 2.13.11 support Java 21. -->
```
Comment (pan3793) [inline pom.xml:1933]: keep our dependencies up to date is an regular work, we don't need this comment

**Comment by pan3793** [comment]
the failure should be fixed by upgrade `maven.plugin.silencer.version` from 1.7.13 to 1.7.17

**Comment by dupen01** [comment]
> the failure should be fixed by upgrade `maven.plugin.silencer.version` from 1.7.13 to 1.7.17

yes, I will upgrade `maven.plugin.silencer.version` to 1.7.17

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
LGTM if CI pass, thanks for your contribution

**Comment by pan3793** [comment]
let's follow https://github.com/apache/spark/pull/46288 to add ignore classes

File: pom.xml [hunk 2]
```diff
@@ -1853,6 +1853,10 @@
                                 <enforceBytecodeVersion>
                                     <maxJdkVersion>${java.version}</maxJdkVersion>
                                     <ignoredScopes>test</ignoredScopes>
+                                    <ignoredScopes>provided</ignoredScopes>
```
Comment (pan3793) [inline pom.xml:1856]:
```suggestion
```

**Comment by pan3793** [comment]
Seems `FinalStageResourceManagerSuite` gets stuck after upgrading Scala 2.13, cc @zhouyifan279 do you have time to take a look? thanks in advance

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/metadata/jdbc/JDBCMetadataStore.scala [hunk 3]
```diff
@@ -80,7 +80,8 @@ class JDBCMetadataStore(conf: KyuubiConf) extends MetadataStore with Logging {
   hikariConfig.setPoolName("jdbc-metadata-store-pool")
 
   @VisibleForTesting
-  implicit private[kyuubi] val hikariDataSource = new HikariDataSource(hikariConfig)
+  implicit private[kyuubi] val hikariDataSource: HikariDataSource =
+    new HikariDataSource(hikariConfig)
```
Comment (bowenliang123) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/metadata/jdbc/JDBCMetadataStore.scala:84]:
```suggestion
  implicit private[kyuubi] val hikariDataSource = new HikariDataSource(hikariConfig)
```
Revert unnecessary change.

File: pom.xml [hunk 4]
```diff
@@ -241,7 +241,7 @@
         <maven.plugin.jacoco.version>0.8.11</maven.plugin.jacoco.version>
         <maven.plugin.scalastyle.version>1.0.0</maven.plugin.scalastyle.version>
         <maven.plugin.shade.version>3.5.2</maven.plugin.shade.version>
-        <maven.plugin.silencer.version>1.7.13</maven.plugin.silencer.version>
+        <maven.plugin.silencer.version>1.7.17</maven.plugin.silencer.version>
```
Comment (bowenliang123) [inline pom.xml:244]: Why we have to update the silencer version?

File: pom.xml [hunk 5]
```diff
@@ -1853,6 +1853,9 @@
                                 <enforceBytecodeVersion>
                                     <maxJdkVersion>${java.version}</maxJdkVersion>
                                     <ignoredScopes>test</ignoredScopes>
+                                    <ignoreClasses>
+                                        <ignoreClass>org.jline.terminal.impl.ffm.*</ignoreClass>
+                                    </ignoreClasses>
```
Comment (bowenliang123) [inline pom.xml:1858]: For what reasons, we have to ignore these classes? Any docs, reference, clarification for this?

File: pom.xml (hunk 4 above)
Comment (pan3793) [reply pom.xml:244]: each version of silencer has a strictly supported Scala version list, it's common to update with Scala version together
https://mvnrepository.com/artifact/com.github.ghik/silencer-plugin

Comment (bowenliang123) [reply pom.xml:244]: Cool. SGTM. Thanks for the explanation.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/metadata/jdbc/JDBCMetadataStore.scala (hunk 3 above)
Comment (dupen01) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/metadata/jdbc/JDBCMetadataStore.scala:84]: With Scala 2.13.11 or above(2.13.9, 2.13.10 not tested), if we use previous code `implicit private[kyuubi] val hikariDataSource = new HikariDataSource(hikariConfig)`,  an error will occur during the compilation phase:

```
kyuubi/kyuubi-server/src/main/scala/org/apache/kyuubi/server/metadata/jdbc/JDBCMetadataStore.scala:83: Implicit definition should have explicit type (inferred com.zaxxer.hikari.HikariDataSource) [quickfixable]
```

File: pom.xml (hunk 5 above)
Comment (dupen01) [reply pom.xml:1858]: As @pan3793 mentioned: https://github.com/apache/spark/pull/46288

Comment (bowenliang123) [reply pom.xml:1858]: OK. But how about bringing the comments as well ? https://github.com/apache/spark/pull/46288/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R2985

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/metadata/jdbc/JDBCMetadataStore.scala (hunk 3 above)
Comment (bowenliang123) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/metadata/jdbc/JDBCMetadataStore.scala:84]: OK.

File: pom.xml (hunk 5 above)
Comment (dupen01) [reply pom.xml:1858]: Sure, I will add the comments.

File: .github/workflows/master.yml [hunk 6]
```diff
@@ -129,18 +129,18 @@ jobs:
...
+    name: Java 17 with Scala 2.13 Test
     runs-on: ubuntu-22.04
     strategy:
       fail-fast: false
       matrix:
         scala:
           - '2.13'
         java:
-          - '8'
+          - '21'
```
Comment (pan3793) [inline .github/workflows/master.yml:141]:
```suggestion
          - '17'
```

**Comment by pan3793** [comment]
I checked the failure message from `FinalStageResourceManagerSuite` 

```
16:21:05.237 ERROR org.apache.spark.network.server.TransportRequestHandler: Error while invoking RpcHandler#receive() on RPC id 9220236832923458123
java.io.InvalidClassException: scala.concurrent.duration.FiniteDuration; local class incompatible: stream classdesc serialVersionUID = -4594686286536372853, local class serialVersionUID = -6513803676778706429
        at java.base/java.io.ObjectStreamClass.initNonProxy(ObjectStreamClass.java:597)
        ...
        at org.apache.spark.serializer.JavaDeserializationStream.readObject(JavaSerializer.scala:87)
        at org.apache.spark.serializer.JavaSerializerInstance.deserialize(JavaSerializer.scala:123)
        at org.apache.spark.rpc.netty.NettyRpcEnv.$anonfun$deserialize$2(NettyRpcEnv.scala:299)
        at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
        at org.apache.spark.rpc.netty.NettyRpcEnv.deserialize(NettyRpcEnv.scala:352)
```

the root cause is that it launches a local cluster using SPARK_HOME, thus using a different Scala library with SPARK_HOME may cause serialization issues as JavaSerializer is not used here.

**Comment by pan3793** [comment]
Thanks, merged to master
